packages/embark-test-runner/src/index.js

```diff
@@ -21,125 +13,73 @@ class TestRunner {
-        const types = { jsFiles: ".js", solidityFiles: "_test.sol" };
-        const groups = Object.entries(types).reduce((acc, [type, ext]) => {
-          acc[type] = files.filter(f => f.endsWith(ext));
-          return acc;
-        }, {});
 
-        next(null, groups);
+    async.waterfall([
+      (next) => {
+        this.getFilesFromDir(testPath, next);
       },
-      (groups, next) => { // run tests
-        let fns = [];
+      (files, next) => {
+        for(const file of files) {
+          const runner = this.runners.find(r => r.matchFn(file));
 
-        if (!options.solc && groups.jsFiles.length > 0) {
-          fns.push((cb) => self.runJSTests(groups.jsFiles, options, cb));
-        } else if (options.solc && groups.solidityFiles.length > 0) {
-          fns.push((cb) => self.runSolidityTests(groups.solidityFiles, options, cb));
-        }
+          if (!runner) {
+            continue;
+          }
 
-        if (fns.length === 0) {
-          return next('No tests to run');
+          runner.addFn(file);
         }
 
-        async.series(fns, next);
+        next();
       },
-      (results, next) => { // generate coverage report
-        if (!options.coverage) {
-          return next(null, results);
-        }
-
-        const cmd = [
-          embarkPath('node_modules/.bin/istanbul'),
-          "report",
-          "--root=.embark",
-          "--format=html",
-          "--format=lcov"
-        ].join(" ");
-
-        runCmd(cmd, {silent: false, exitOnError: false}, (err) => {
-          if (err) {
-            return next(err);
-          }
-
-          self.logger.info(`Coverage report created. You can find it here: ${dappPath('coverage/index.html')}`);
-
-          if (options.noBrowser) {
-            return next(null, results);
-          }
-
-          const opn = require('opn');
-          const _next = () => { next(null, results); };
-
-          opn(dappPath('coverage/index.html'), {wait: false})
-            .then(() => timer(1000))
-            .then(_next, _next);
+      (next) => {
+        reporter.header();
 
+        options.reporter = reporter;
+        const runnerFns = this.runners.map((runner) => {
+          return (_cb) => { runner.runFn(options, _cb); };
         });
-      },
-      (results, next) => { // show report
-        const totalFailures = results.reduce((acc, result) => acc + result, 0);
 
-        (totalFailures == 0)
-          ? next(null, ' > All tests passed'.green.bold)
-          : next(totalFailures, ` > Total number of failures: ${totalFailures}`.red.bold);
+        async.series(runnerFns, next);
       }
-    ], (err, msg) => {
-      process.stdout.write(msg + "\n");
-
-      self.fs.remove('.embark/contracts');
-      self.fs.remove('.embark/remix_tests.sol');
-
-      return cb(err);
+    ], (err) => {
+      reporter.footer();
+      cb(err, reporter.passes, reporter.fails);
     });
   }
 
-
   getFilesFromDir(filePath, cb) {
```

Comment:
Can just return the promise from `fs.stat` to make this `async` and get rid of the waterfall.